src/error.rs

```diff
@@ -33,6 +33,37 @@ pub extern "C" fn rustls_error(
     }
 }
 
+#[no_mangle]
+pub extern "C" fn rustls_result_is_cert_error(result: rustls_result) -> bool {
+    match result {
+        rustls_result::CertBadDER => true,
+        rustls_result::CertBadDERTime => true,
+        rustls_result::CertCAUsedAsEndEntity => true,
+        rustls_result::CertExpired => true,
+        rustls_result::CertNotValidForName => true,
+        rustls_result::CertNotValidYet => true,
+        rustls_result::CertEndEntityUsedAsCA => true,
+        rustls_result::CertExtensionValueInvalid => true,
+        rustls_result::CertInvalidCertValidity => true,
+        rustls_result::CertInvalidSignatureForPublicKey => true,
+        rustls_result::CertNameConstraintViolation => true,
+        rustls_result::CertPathLenConstraintViolated => true,
+        rustls_result::CertSignatureAlgorithmMismatch => true,
+        rustls_result::CertRequiredEKUNotFound => true,
+        rustls_result::CertUnknownIssuer => true,
+        rustls_result::CertUnsupportedCertVersion => true,
+        rustls_result::CertUnsupportedCriticalExtension => true,
+        rustls_result::CertUnsupportedSignatureAlgorithmForPublicKey => true,
+        rustls_result::CertUnsupportedSignatureAlgorithm => true,
+        rustls_result::CertSCTMalformed => true,
+        rustls_result::CertSCTInvalidSignature => true,
+        rustls_result::CertSCTTimestampInFuture => true,
+        rustls_result::CertSCTUnsupportedVersion => true,
+        rustls_result::CertSCTUnknownLog => true,
+        _ => false,
+    }
+}
+
```

Comment:
It's a shame that Rust doesn't let you do something like Java enums do, where you can embed values in an enum case, so you could do something like:

```
pub enum rustls_result {
  Io(false),
  CertBadDER(true),
  ...
};
```
...where the boolean value indicates whether the enclosing enum case is a cert error (you could find a more expressive way to encode that than a bool, of course). I just think you might end up with a lot of long functions like this one as you add more attributes to these enum cases, and it's easy to end up with bugs resulting from forgetting to list all the appropriate enum cases in a new `rustls_result_is_<whatever>` function.

Reply:
Well, the Rust does of course allow you to embed values in enums. In your example it would be:

```rust
pub enum rustls_result {
  Io(bool),
  CertBadDER(bool),
}
```

And in fact [TLSError uses these embedded values a lot](https://docs.rs/rustls/0.19.0/rustls/enum.TLSError.html). But `rustls_result` is the C mapping of that enum, so I intentionally flattened all the enums-with-values into one namespace.

The more principled way to do this would be to define a `reverse_map_error` that takes a `rustls_result`, and turns it into a `TLSError`. Then `rustls_is_cert_error` would check if it was the variant `WebPKIError` or `InvalidSCT`.

To some extent, `result_to_string` is this reverse mapping, except instead of mapping into `TLSError` it maps into `String`. I did try initially to write the reverse mapping, but ran into some troubles. Maybe I should give it another try. You're right that giant lists invite bugs due to omitting something.

Reply:
Alright, tried it with the reverse mapping. Let me know what you think!